src/enzo/ActiveParticleRoutines.C

```diff
@@ -0,0 +1,508 @@
+  DynamicalTime = part->DynamicalTime;
+  Metallicity = part->Metallicity;
+  Identifier = part->Identifier;
+  level = part->level;
+  GridID = part->GridID;
+  type = part->type;
+  CurrentGrid = part->CurrentGrid;
+  WillDelete = part->WillDelete;
+}
+
+ActiveParticleType::ActiveParticleType(grid *_grid, ActiveParticleFormationData &data)
+{
+  int dim;
+  CurrentGrid = _grid;
+  for (dim = 0; dim < MAX_DIMENSION; dim++)
+    pos[dim] = vel[dim] = 0.0;
+  Mass = BirthTime = DynamicalTime = 0.0;
+  type = 0;
+  level = data.level;
+  GridID = data.GridID;
+  /* The correct indices are assigned in CommunicationUpdateActiveParticleCount 
+     in ActiveParticleFinalize.*/
+  Identifier = INT_UNDEFINED;
+
+  // Assume newly created particles will not be immediately deleted
+  WillDelete = false;
+}
+
+
+ActiveParticleType::~ActiveParticleType(void)
+{
+}
+
+/***************
+
+    OPERATORS
+
+ ***************/
+
+void ActiveParticleType::operator=(ActiveParticleType *a)
+{
+  int i, dim;
+  CurrentGrid = a->CurrentGrid;
+  for (dim = 0; dim < MAX_DIMENSION; dim++) {
+    pos[dim] = a->pos[dim];
+    vel[dim] = a->vel[dim];
+  }
+  Mass = a->Mass;
+  BirthTime = a->BirthTime;
+  DynamicalTime = a->DynamicalTime;
+  Metallicity = a->Metallicity;
+  Identifier = a->Identifier;
+  level = a->level;
+  GridID = a->GridID;
+  type = a->type;
+  return;
+}
+
+/**********************
+
+   CONVENIENT ROUTINES
+
+ **********************/
+
+template<class active_particle_class>
+active_particle_class *ActiveParticleType::copy(void)
+{
+  int i, dim;
+  active_particle_class *a = new active_particle_class();
+  a->CurrentGrid = CurrentGrid;
+  for (dim = 0; dim < MAX_DIMENSION; dim++) {
+    a->pos[dim] = pos[dim];
+    a->vel[dim] = vel[dim];
+  }
+  a->Mass = Mass;
+  a->BirthTime = BirthTime;
+  a->DynamicalTime = DynamicalTime;
+  a->Metallicity = Metallicity;
+  a->Identifier = Identifier;
+  a->level = level;
+  a->GridID = GridID;
+  a->type = type;
+  return a;
+}
+
+void ActiveParticleType::ConvertAllMassesToSolar(void)
+{
+  const double Msun = 1.989e33;
+  double dx;
+  float DensityUnits, LengthUnits, TemperatureUnits, TimeUnits,
+    VelocityUnits, MassConversion;
+  GetUnits(&DensityUnits, &LengthUnits, &TemperatureUnits,
+	   &TimeUnits, &VelocityUnits, CurrentGrid->Time);
+  dx = LengthUnits * CurrentGrid->CellWidth[0][0];
+  MassConversion = (float) (dx*dx*dx * double(DensityUnits) / Msun);
+  this->Mass *= MassConversion;
+  return;
+}
+
+void ActiveParticleType::ConvertMassToSolar(void)
+{
+  const double Msun = 1.989e33;
+  double dx;
+  float DensityUnits, LengthUnits, TemperatureUnits, TimeUnits,
+    VelocityUnits, MassConversion;
+  GetUnits(&DensityUnits, &LengthUnits, &TemperatureUnits,
+	   &TimeUnits, &VelocityUnits, CurrentGrid->Time);
+  dx = LengthUnits * CurrentGrid->CellWidth[0][0];
+  MassConversion = (float) (dx*dx*dx * double(DensityUnits) / Msun);
+  this->Mass *= MassConversion;
+  return;
+}
+
```

Comment:
I think function 'ConvertAllMassesToSolar' and 'ConvertMassToSolar' are identical. Is both required?

Reply:
This is redundant. I'll remove the "AllMasses" function and make the appropriate changes through the code.